### PR TITLE
Evolve BrokerCapacityConfigResolver interface to be aware whether capacity estimation is allowed or not in broker capacity resolution.

### PR DESCRIPTION
**Pull request by kidkun**
Addresses the issue #1107 .
The main change in this PR to evolve [BrokerCapacityConfigResolver#capacityForBroker](https://github.com/linkedin/cruise-control/blob/c84af32fcc96d1c4d026e8e4dd5aac54bb8ace6f/cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/config/BrokerCapacityConfigResolver.java#L38) method signature by adding a new argument `allowCapacityEstimation `. 

Doing so have multiple advantages. 
 1. Now `BrokerCapacityConfigResolver` is aware that estimation is allowed or not, so it no longer makes blindly assumption.
2. Now in cluster model generation process, if `BrokerCapacityConfigResolver`  returns failing response, we can fail early; if `BrokerCapacityConfigResolver`  returns success response, we are certain that the generated cluster model satisfies estimation requirement so sanity check can be avoided.

### CODE REVIEW

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/config/BrokerCapacityConfigResolver.java [hunk 1]
```diff
@@ -30,8 +30,11 @@
    * @param host The host of the broker
    * @param brokerId The id of the broker
    * @param timeoutMs The timeout in millisecond.
+   * @param allowCapacityEstimation Whether allow resolver to estimate broker capacity if resolver is unable to get
```
[the remark this answers is not in the document]
Comment (kidkun) [reply cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/config/BrokerCapacityConfigResolver.java]: Discuss offline, we will create a new `BrokerCapacityResolvingException` type to represent capacity resolving failure.

File: cruise-control/src/test/java/com/linkedin/kafka/cruisecontrol/detector/AnomalyDetectorTest.java [hunk 2]
```diff
@@ -293,7 +293,7 @@ private void testFixAnomaly(AnomalyType anomalyType)
       EasyMock.expect(mockAnomalyNotifier.onGoalViolation(EasyMock.isA(GoalViolations.class))).andReturn(AnomalyNotificationResult.fix());
     } else if (anomalyType == KafkaAnomalyType.DISK_FAILURE) {
       ClusterModel singleBrokerWithBadDisk = singleBrokerWithBadDisk();
-      EasyMock.expect(mockKafkaCruiseControl.clusterModel(EasyMock.anyObject(), EasyMock.anyObject())).andReturn(singleBrokerWithBadDisk);
+      EasyMock.expect(mockKafkaCruiseControl.clusterModel(EasyMock.anyObject(), EasyMock.anyObject(), EasyMock.anyObject())).andReturn(singleBrokerWithBadDisk);
```
[the remark this answers is not in the document]
Comment (kidkun) [reply cruise-control/src/test/java/com/linkedin/kafka/cruisecontrol/detector/AnomalyDetectorTest.java]: I feel these scenario is specific to `LoadMonitor`, I prefer to just check in `LoadMonitorTest`, this test should focus on how detector react to different anomalies.

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/config/BrokerCapacityConfigFileResolver.java [hunk 3]
```diff
@@ -178,8 +179,9 @@ public BrokerCapacityInfo capacityForBroker(String rack, String host, int broker
                                         _capacitiesForBrokers.get(DEFAULT_CAPACITY_BROKER_ID).diskCapacityByLogDir(),
                                         _capacitiesForBrokers.get(DEFAULT_CAPACITY_BROKER_ID).numCpuCores());
         } else {
-          throw new TimeoutException(String.format("Unable to resolve capacity of broker %d. Either allow capacity estimation "
-                                                   + "or add broker's capacity information in file %s.", brokerId, _configFile));
+          throw new BrokerCapacityResolvingException(String.format("Unable to resolve capacity of broker %d.Either (1) adding the "
```
Comment (efeg) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/config/BrokerCapacityConfigFileResolver.java]: Nit: `%d.Either` -> `%d. Either`

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/config/BrokerCapacityConfigFileResolver.java [hunk 4]
```diff
@@ -178,8 +179,9 @@ public BrokerCapacityInfo capacityForBroker(String rack, String host, int broker
...
+              + "default broker capacity (via adding capacity for broker %d and allowing capacity estimation or (2) add missing "
```
Comment (efeg) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/config/BrokerCapacityConfigFileResolver.java]: Missing closing parenthesis.

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/config/BrokerCapacityConfigFileResolver.java (hunk 3 above)
Comment (efeg) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/config/BrokerCapacityConfigFileResolver.java]: Inconsistent use of `add` and `adding` 
should we reformat it as -> `Either add (1) XXX, or (2) YYY.`

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/config/BrokerCapacityConfigResolver.java [hunk 5]
```diff
@@ -32,9 +25,10 @@
    * @param timeoutMs The timeout in millisecond.
    * @param allowCapacityEstimation Whether allow resolver to estimate broker capacity if resolver is unable to get
    *                                capacity information of the broker.
-   * @return The capacity of each resource for the broker
+   * @return A instance of {@link BrokerCapacityInfo}.
```
Comment (efeg) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/config/BrokerCapacityConfigResolver.java]: Nit: -> `An instance`

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/config/BrokerCapacityInfo.java [hunk 6]
```diff
@@ -10,6 +10,17 @@
 import java.util.stream.Collectors;
 
 
+/**
+ * A class encapsulate the capacity information of a broker, which includes all the resources defined in {@link Resource}.
```
Comment (efeg) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/config/BrokerCapacityInfo.java]: Nit: `encapsulate` -> `encapsulating`

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/exception/BrokerCapacityResolvingException.java [hunk 7]
```diff
@@ -0,0 +1,19 @@
+/*
+ * Copyright 2020 LinkedIn Corp. Licensed under the BSD 2-Clause License (the "License"). See License in the project root for license information.
+ */
+
+package com.linkedin.kafka.cruisecontrol.exception;
+
+/**
+ * The exception indicating something went wrong during the broker capacity resolving.
+ */
+public class BrokerCapacityResolvingException extends KafkaCruiseControlException {
```
Comment (efeg) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/exception/BrokerCapacityResolvingException.java]: Nit: `BrokerCapacityResolvingException` -> `CapacityResolutionException` or `BrokerCapacityResolutionException`

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/LoadMonitor.java [hunk 8]
```diff
@@ -506,32 +510,34 @@ public ClusterModel clusterModel(long from,
...
+      // If broker capacity is allowed to estimate broker capacity, shuffle nodes before getting their capacity from the
+      // capacity resolver. This is good for the capacity resolver to estimate the capacity of the nodes, for which the
+      // capacity retrieval has failed.
       // The use case for this estimation is that if the capacity of one of the nodes is not available (e.g. due to some
       // 3rd party service issue), the capacity resolver may want to use the capacity of a peer node as the capacity for
       // that node.
       // To this end, Cruise Control handles the case that the first node is problematic so the capacity resolver does
       // not have the chance to get the capacity for the other nodes.
       // Shuffling the node order helps, as the problematic node is unlikely to always be the first node in the list.
       List<Node> shuffledNodes = new ArrayList<>(cluster.nodes());
```
Comment (efeg) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/LoadMonitor.java]: Do we even need to create a new `ArrayList` in case `allowCapacityEstimation` is false?

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/config/BrokerCapacityConfigFileResolver.java [hunk 9]
```diff
@@ -156,17 +156,18 @@
 
   @Override
   public void configure(Map<String, ?> configs) {
-    _configFile = KafkaCruiseControlUtils.getRequiredConfig(configs, CAPACITY_CONFIG_FILE);
+    String configFile = KafkaCruiseControlUtils.getRequiredConfig(configs, CAPACITY_CONFIG_FILE);
+    _configFile = configFile;
     try {
-      loadCapacities(_configFile);
+      loadCapacities(configFile);
```
Comment (efeg) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/config/BrokerCapacityConfigFileResolver.java]: Sorry if my previous comment was not clear -- let me rephrase my question:

* Why are we passing a parameter to `loadCapacities` function? This is a local private function and we already have the value assigned to `_configFile`. Is it possible to modify `loadCapacities` to have no parameter and use `_configFile` in its implementation?

File: cruise-control/src/test/java/com/linkedin/kafka/cruisecontrol/config/BrokerCapacityConfigFileResolverTest.java [hunk 10]
```diff
@@ -48,10 +49,13 @@ public void testParseConfigFile() throws TimeoutException {
 
     assertTrue(configResolver.capacityForBroker("", "", 2, BROKER_CAPACITY_FETCH_TIMEOUT_MS, true).isEstimated());
     assertTrue(configResolver.capacityForBroker("", "", 2, BROKER_CAPACITY_FETCH_TIMEOUT_MS, true).estimationInfo().length() > 0);
+
+    // If resolver is unable to get broker capacity and not allowed to estimate, an BrokerCapacityResolvingException will be thrown.
```
Comment (efeg) [inline cruise-control/src/test/java/com/linkedin/kafka/cruisecontrol/config/BrokerCapacityConfigFileResolverTest.java]: Nit: `an BrokerCapacityResolvingException` -> `a BrokerCapacityResolvingException`

File: cruise-control/src/test/resources/DefaultCapacityConfig.json [hunk 11]
```diff
@@ -11,14 +11,14 @@
...
     {
-      "brokerId": "314",
+      "brokerId": "1",
       "capacity": {
         "DISK": "150000",
         "CPU": "100",
         "NW_IN": "150000",
         "NW_OUT": "150000"
       },
-      "doc": "This overrides the capacity for broker 314."
```
Comment (efeg) [inline cruise-control/src/test/resources/DefaultCapacityConfig.json]: Nit: I am curious why this change is needed.

Comment (kidkun) [reply cruise-control/src/test/resources/DefaultCapacityConfig.json]: This is because in many CC tests we use cluster model with broker id 0 and 1 in it.
 So all brokers are used to be estimated, now there will be one estimated and one accurate, the tests actually cover more scenarios automatically.
